std/algorithm/searching.d

```diff
@@ -4192,9 +4192,24 @@ bool startsWith(alias pred = "a == b", R, E)(R doesThisStart, E withThis)
 if (isInputRange!R &&
     is(typeof(binaryFun!pred(doesThisStart.front, withThis)) : bool))
 {
-    return doesThisStart.empty
-        ? false
-        : binaryFun!pred(doesThisStart.front, withThis);
+    if (doesThisStart.empty)
+        return false;
+
+    alias predFunc = binaryFun!pred;
+
+    // auto-decoding special case
+    static if (isNarrowString!R)
+    {
+        // specialize for ASCII as to not change previous behavior
+        if (withThis <= 0x7F)
+            return predFunc(doesThisStart[0], withThis);
```

Comment:
Not really about "why would you want" as it is changing behaviour in general in an optimization PR.

At least for `E==char`, I can see two possibilities:

1. `needle` is a Unicode code point crammed into an UTF-8 code unit, which actually works for a certain code page (Latin 1, was it?).

2. `needle` is an UTF-8 code unit (not necessarily part of a Latin-1 character), and `haystack` is doubly-encoded (an UTF-8 string, where each `char` is interpreted as a character and then again encoded to UTF-8).

Both of these are pretty broken, but nevertheless they are two cases where removing that `if` check would result in changing behaviour.